lib/file-source/src/paths_provider/glob.rs

```diff
@@ -14,54 +42,83 @@ pub use glob::MatchOptions;
     pub fn new(
-        include_patterns: &[PathBuf],
-        exclude_patterns: &[PathBuf],
-        glob_match_options: MatchOptions,
-    ) -> Option<Self> {
-        let include_patterns = include_patterns
+        include_patterns: &[String],
+        exclude_patterns: &[String],
+    ) -> Result<Self, GlobError> {
+        // Validate include patterns. We can't parse the `GlobWalkers` and save them in our struct
+        // for later use because they are mutable iterators and don't implement the
+        // `std::clone::Clone` trait.
+        include_patterns
             .iter()
-            .map(|path| path.to_str().map(ToOwned::to_owned))
-            .collect::<Option<_>>()?;
+            .map(|include_pattern| {
+                glob(include_pattern).with_context(|| InvalidIncludePattern {
+                    glob: include_pattern.to_owned(),
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
+        let include_patterns = include_patterns.to_owned();
+
+        // Validate exclude patterns.
         let exclude_patterns = exclude_patterns
             .iter()
-            .map(|path| path.to_str().map(|path| Pattern::new(path).ok()))
-            .flatten()
-            .collect::<Option<Vec<_>>>()?;
+            .map(|exclude_pattern| -> Result<_, _> {
+                let pattern =
+                    Pattern::new(exclude_pattern).with_context(|| InvalidExcludePattern {
+                        glob: exclude_pattern.to_owned(),
+                    })?;
+
+                Ok(pattern)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
-        Some(Self {
+        Ok(Self {
             include_patterns,
             exclude_patterns,
-            glob_match_options,
         })
     }
 }
 
 impl PathsProvider for Glob {
     type IntoIter = Vec<PathBuf>;
+    type Error = GlobError;
 
-    fn paths(&self) -> Self::IntoIter {
-        self.include_patterns
-            .iter()
-            .flat_map(|include_pattern| {
-                glob::glob_with(include_pattern.as_str(), self.glob_match_options)
-                    .expect("failed to read glob pattern")
-                    .filter_map(|val| val.ok())
-            })
-            .filter(|candidate_path: &PathBuf| -> bool {
-                !self.exclude_patterns.iter().any(|exclude_pattern| {
-                    let candidate_path_str = candidate_path.to_str().unwrap();
-                    exclude_pattern.matches(candidate_path_str)
-                })
-            })
-            .collect()
+    fn paths(&self) -> Result<Self::IntoIter, Self::Error> {
+        let mut paths = Vec::new();
+
+        // Iterate over all include patterns, turn them into globs and walk them on the file system.
+        for include_pattern in &self.include_patterns {
+            let glob = glob(include_pattern).with_context(|| InvalidIncludePattern {
+                glob: include_pattern.to_owned(),
+            })?;
```

Comment:
In relation to https://github.com/timberio/vector/pull/5927/files#r559039229, and since we're validating that the include patterns are valid at construction, it might be meaningful to assume they're valid here and elide the possibility of an error via an `unwrap`.
If I understand correctly, once we've validated the patterns above, it is not possible that they will actually return errors here?

Reply:
Agree, either unwrapping or parsing into a more approriate type would be the ideal behavior here.

Let's see what we can do here after the changes to `glob` or `globwalk`.